lib/charms/observability_libs/v0/juju_topology.py

```diff
@@ -0,0 +1,230 @@
+                - "unit"
+                - "charm_name"
+                `unit` and `charm_name` may be empty, but will result in more limited
+                labels. However, this allows us to support payload-only charms.
+
+        Returns:
+            a `JujuTopology` object.
+        """
+        return cls(
+            model=data["model"],
+            model_uuid=data["model_uuid"],
+            application=data["application"],
+            unit=data.get("unit", ""),
+            charm_name=data.get("charm_name", ""),
+        )
+
+    def as_dict(
+        self,
+        *,
+        remapped_keys: Dict[str, str] = {},
+        excluded_keys: List[str] = [],
+        uuid_length: Optional[int] = None,
+    ) -> OrderedDict:
+        """Format the topology information into an ordered dict.
+
+        Keeping the dictionary ordered is important to be able to
+        compare dicts without having to resort to deep comparisons.
+
+        Args:
+            remapped_keys: A dictionary mapping old key names to new key names,
+                which will be substituted when invoked.
+            excluded_keys: A list of key names to exclude from the returned dict.
+            uuid_length: The length to crop the UUID to.
+        """
+        model_uuid = self.model_uuid[: uuid_length - 1] if uuid_length else self.model_uuid
+        ret = OrderedDict(
+            [
+                ("model", self.model),
+                ("model_uuid", model_uuid),
+                ("application", self.application),
+                ("unit", self.unit),
+                ("charm_name", self.charm_name),
+            ]
+        )
+
+        for exclusion in excluded_keys:
+            if exclusion in ret:
+                ret.pop(exclusion)
+
+        if remapped_keys:
+            ret = OrderedDict(
+                (remapped_keys.get(k), v) if remapped_keys.get(k) else (k, v) for k, v in ret.items()  # type: ignore
+            )
+
+        return ret
+
+    def as_identifier(self) -> str:
+        """Format the topology information into a terse string.
+
+        This crops the model UUID, making it not suitable for comparisons against
+        anything but other identifiers.
+        """
+        values = self.as_dict(
+            excluded_keys=["unit", "charm_name"],
+            uuid_length=8,
+        ).values()
+
+        return "_".join([str(val) for val in values]).replace("/", "_")
+
+    def as_relabelled_dict(self):
+        """Format the topology information into a dict with keys having 'juju_' as prefix.
+
+        Relabelled topology never includes the unit as it would then only match
+        the leader unit (ie. the unit that produced the dict).
+        """
+        items = self.as_dict(
+            remapped_keys={"charm_name": "charm"},
+            excluded_keys=["unit"],
+        ).items()
+
+        return {"juju_{}".format(key): value for key, value in items if value}
+
+    def as_label_matchers(self) -> str:
```

Comment:
I would drop this function in favor of using `app_identifier` with a new additional arg `separator=', '`

Reply:
I prefer readability. It's not like the methods are doing much, so we might as well keep it. Not a super fan of adding additional arguments unless it is unavoidable.

Reply:
I am also in favour of dropping this code and instead ensuring that `from_charm()` either
- always includes unit (the unit that constructed it)
- never includes unit (user of class has to add it if they need to)
Juju Topology is doing too many things and trying to satisfy too many use cases. This makes its interface awkward and difficult to use.

Reply:
I disagree. It currently has one method and 3 properly documented properties. The reason why we don't want to always have it prefixed with `juju_` is simple: that is not how the charm itself sees it, so then we'd have no way of reconstructing the original data fed to the factory or constructor.